var/spack/repos/builtin/packages/daos/package.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class Daos(SConsPackage):
+    """The Distributed Asynchronous Object Storage (DAOS) is an open-source
+    software-defined object store designed from the ground up for massively
+    distributed Non Volatile Memory (NVM)."""
+
+    homepage = "https://github.com/daos-stack/daos"
+    git = "https://github.com/daos-stack/daos.git"
+    maintainers("hyoklee")
+
+    version("master", branch="master", submodules=True)
+    version("2.2.0", tag="v2.2.0", submodules=True)
+    variant(
+        "debug", default=False, description="Enable debugging info and strict compile warnings"
+    )
+
+    depends_on("argobots")
+    depends_on("boost@develop+python", type="build")
+    depends_on("cmocka", type="build")
+    depends_on("dpdk@main")
+    depends_on("go", type="build")
+    depends_on("hwloc")
+    depends_on("isa-l")
+    depends_on("isa-l_crypto")
+    depends_on("libfabric")
+    depends_on("libfuse@3.6.1:")
+    depends_on("libuuid")
+    depends_on("libunwind")
+    depends_on("libyaml")
+    depends_on("mercury+boostsys")
+    depends_on("mpich")
+    depends_on("openssl")
+    depends_on("pmdk")
+    depends_on("protobuf-c")
+    depends_on("py-distro")
+    depends_on("readline")
+    depends_on("scons@4.4.0:")
+    depends_on("spdk+shared+rdma")
+
+    def build_args(self, spec, prefix):
+        args = ["PREFIX={0}".format(prefix), "--build-deps=yes"]
```

Comment:
you need to remove --build-deps=yes here otherwise that would tell daos to build itself all the dependencies

Reply:
@soumagne , thank you for your review.

Unfortunately, this is required due to DAOS rdb git submodule:

https://github.com/daos-stack/daos/tree/master/src/rdb

There's no Spack package for DAOS rdb. 

Do you have any idea on how to handle this?
I can think of 2 options:

1. Create a new daos-rdb package and use it as a dependent.
2. Add +build_deps variant.

Please advise!

Reply:
Yes. I get an error without it.

Reply:
I'm using master because of https://github.com/daos-stack/daos/pull/11393.

Reply:
ok thanks I will take a look

Reply:
I don't see any issues with building master or 2.2.0 without passing `--build-deps=yes` so maybe try again or please send me the error log that you get. I would suspect that enabling `--build-deps` is potentially hiding a dependency issue that you have on your end. In any case, it should be disabled.

Reply:
See https://github.com/hyoklee/spack/actions/runs/4297917161 for the failure.

Here's the error message:

```
2023-03-01T00:45:57.3800340Z BuildRequired: ucx needs to be built, use --build-deps=yes:
2023-03-01T00:45:57.4064304Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/SConstruct", line 501:
2023-03-01T00:45:57.4065058Z     scons()
2023-03-01T00:45:57.4066145Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/SConstruct", line 440:
2023-03-01T00:45:57.4066823Z     prereqs.run_build(opts)
2023-03-01T00:45:57.4067926Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/site_scons/prereq_tools/base.py", line 559:
2023-03-01T00:45:57.4068626Z     self.__env.Clone().require(comp)
2023-03-01T00:45:57.4069796Z   File "/home/runner/work/spack/spack/opt/spack/linux-ubuntu22.04-x86_64_v4/gcc-12.1.0/scons-4.4.0-btcyyrncqhd4g2rnail7oikfbj4hctef/lib/python3.10/site-packages/SCons/Util.py", line 737:
2023-03-01T00:45:57.4070502Z     return self.method(*nargs, **kwargs)
2023-03-01T00:45:57.4071587Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/site_scons/prereq_tools/base.py", line 807:
2023-03-01T00:45:57.4072252Z     raise error
2023-03-01T00:45:57.4073311Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/site_scons/prereq_tools/base.py", line 797:
2023-03-01T00:45:57.4074332Z     if comp_def.build(env, needed_libs):
2023-03-01T00:45:57.4075434Z   File "/tmp/runner/spack-stage/spack-stage-daos-master-yvkyuhtzz2z4bz2wmeivkwzdnwf7vs2q/spack-src/site_scons/prereq_tools/base.py", line 1407:
2023-03-01T00:45:57.4076133Z     raise BuildRequired(self.name)
2023-03-01T00:45:57.4360710Z ==> Error: ProcessError: Command exited with status 2:
```

Reply:
ah thanks a lot for sending that log. Yes you're missing a dependency on `ucx` in that case (there's already a package for it in spack), eventually we'd need to add a variant for it but I don't think that's possible right now.

Reply:
Yes, although I include ucx as dependent, Scons doesn't pick it up.

https://github.com/hyoklee/spack/actions/runs/4298976035

I mentioned https://github.com/spack/spack/issues/4927 in the first comment of this PR.
I've spent a lot of time and tried many hacks but Spack and Scons don't get along well.

Thus, I wish Intel DAOS team uses CMake only and remove Scons for DAOS as a build system.

Reply:
:) I see, don't worry I'm pretty sure Scons can pick it up. Indeed scons does not use spack compiler wrappers but instead packages are passed through the `ALT_PREFIX` so you would need to add to add the ucx dependency and pass `format(spec["ucx"].prefix),` also to the ALT_PREFIX so that it gets picked up there as well. This will then tell scons about the ucx install and where to look for ucx libraries etc.

Reply:
Right, I keep forgetting about `ALT_PREFIX`!

Reply:
Yes it's not perfect but it should work. Also you might need to pass the flag `USE_INSTALLED=all` to the args.

Reply:
`hwloc` fails although I specify it in ALT_PREFIX.

```
Checking targets for component 'hwloc'
Checking for C header file hwloc.h... no
MissingTargets: Package hwloc-devel is required. Check config.log:
  File "/tmp/hyoklee/spack-stage/spack-stage-daos-master-kpcfuqimmjkxzzebduxpr7dh\
zaxfgveu/spack-src/SConstruct", line 501:
    scons()
```

Reply:
could be a bug that needs to be fixed in the daos scons files